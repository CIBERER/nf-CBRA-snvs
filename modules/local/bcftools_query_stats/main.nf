process BCFTOOLS_QUERY_STATS {
    tag "$meta.id"
    label 'process_low'

    conda "bioconda::bcftools=1.15.1"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bcftools:1.15.1--h0ea216a_0' :
        'quay.io/biocontainers/bcftools:1.15.1--h0ea216a_0' }"

    input:
    tuple val(meta), path(vcf), path(tbi)

    output:
    tuple val(meta), path("*_GT.txt"), emit: gt
    tuple val(meta), path("*_DP.txt"), emit: dp
    tuple val(meta), path("*_VD.txt"), emit: vd
    tuple val(meta), path("*_RD.txt"), emit: rd
    tuple val(meta), path("*_DP_mean.txt"), emit: dp_mean
    tuple val(meta), path("*_VD_mean.txt"), emit: vd_mean
    tuple val(meta), path("*_RD_mean.txt"), emit: rd_mean
    tuple val(meta), path("*_AD_mean.txt"), emit: ad_mean
    tuple val(meta), path("*_VAF.txt"), emit: vaf
    tuple val(meta), path("*_header.txt"), emit: programs
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Extract sample names from the VCF and split to get program names
    programs=\$(bcftools query -l ${vcf} | sed 's/.*\\.//' | tr '\n' '\t' | sed 's/\\t\$//')

    echo -e "\$programs" > ${prefix}_header.txt

    # Get genotype, depth and variant allele depth
    bcftools query -f '[\\t%GT]\\n' ${vcf} | sed 's/\\t//1' | sed 's/\\.\\/\\.//g' | sed 's/1\\/0/0\\/1/g' > ${prefix}_GT_body.txt
    cat ${prefix}_header.txt ${prefix}_GT_body.txt > ${prefix}_GT.txt

    bcftools query -f '[\\t%DP]\\n' ${vcf} | sed 's/\\t//1' | sed 's/\\.//g' > ${prefix}_DP.txt
    bcftools query -f '[\\t%AD{1}]\\n' ${vcf} | sed 's/\\t//1' | sed 's/\\.//g' > ${prefix}_VD.txt
    bcftools query -f '[\\t%AD{0}]\\n' ${vcf} | sed 's/\\t//1' | sed 's/\\.//g' > ${prefix}_RD.txt

    # Calculate the average depth and variant allele depth
    awk -v OFMT=%.0f '{sum = 0; for (i = 1; i <= NF; i++) sum += \$i; sum /= NF; print sum}' ${prefix}_DP.txt > ${prefix}_DP_mean.txt
    awk -v OFMT=%.0f '{sum = 0; for (i = 1; i <= NF; i++) sum += \$i; sum /= NF; print sum}' ${prefix}_VD.txt > ${prefix}_VD_mean.txt
    awk -v OFMT=%.0f '{sum = 0; for (i = 1; i <= NF; i++) sum += \$i; sum /= NF; print sum}' ${prefix}_RD.txt > ${prefix}_RD_mean.txt
    paste -d, ${prefix}_RD_mean.txt ${prefix}_VD_mean.txt > ${prefix}_AD_mean.txt

    # Calculate variant allele depth (VAD)
    paste ${prefix}_VD_mean.txt ${prefix}_DP_mean.txt | awk -v OFMT=%.2f '{if (\$2 == 0) print "-nan"; else print(\$1/\$2)}' > ${prefix}_VAF.txt


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """
}