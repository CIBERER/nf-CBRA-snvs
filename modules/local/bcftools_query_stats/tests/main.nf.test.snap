{
    "bcftools_query_stats - vcf ": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_GT.txt:md5,c6bb82e3fc42c6049b131e990cce7431"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_DP.txt:md5,5cf0e6d081a47f175cd4756789455b2c"
                    ]
                ],
                "10": [
                    "versions.yml:md5,8c82948b5ba327ba4175106003d21255"
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_VD.txt:md5,313ae63a1dc071604f6bf8f1e6dfefaa"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_RD.txt:md5,ae59a3ee19f9a594321a4e2752033ee8"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_DP_mean.txt:md5,0add7edd323f7a7df46c1118870f9dd9"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_VD_mean.txt:md5,58351ef788f43dcb71a58d8bd847c688"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_RD_mean.txt:md5,0c89c4c336ea87db4775310f5c73f263"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_AD_mean.txt:md5,a38ef9a1461a609010a13d56112b4cfb"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_VAF.txt:md5,6d7393b63d68c9163de8d99a1ea3bc91"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_header.txt:md5,d694a13ad4d64e9a0a455f3e3e29c150"
                    ]
                ],
                "ad_mean": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_AD_mean.txt:md5,a38ef9a1461a609010a13d56112b4cfb"
                    ]
                ],
                "dp": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_DP.txt:md5,5cf0e6d081a47f175cd4756789455b2c"
                    ]
                ],
                "dp_mean": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_DP_mean.txt:md5,0add7edd323f7a7df46c1118870f9dd9"
                    ]
                ],
                "gt": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_GT.txt:md5,c6bb82e3fc42c6049b131e990cce7431"
                    ]
                ],
                "programs": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_header.txt:md5,d694a13ad4d64e9a0a455f3e3e29c150"
                    ]
                ],
                "rd": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_RD.txt:md5,ae59a3ee19f9a594321a4e2752033ee8"
                    ]
                ],
                "rd_mean": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_RD_mean.txt:md5,0c89c4c336ea87db4775310f5c73f263"
                    ]
                ],
                "vaf": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_VAF.txt:md5,6d7393b63d68c9163de8d99a1ea3bc91"
                    ]
                ],
                "vd": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_VD.txt:md5,313ae63a1dc071604f6bf8f1e6dfefaa"
                    ]
                ],
                "vd_mean": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_VD_mean.txt:md5,58351ef788f43dcb71a58d8bd847c688"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8c82948b5ba327ba4175106003d21255"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-06-23T08:27:00.385688885"
    }
}